kv/tikv/server.go

```diff
@@ -1,140 +1,85 @@
-		defer server.Latches.ReleaseLatches(latches)
-
-		reader, err := server.innerServer.Reader(ctxt)
-		if err != nil {
-			return nil, err
-		}
-		defer reader.Close()
-
-		// Build an mvcc transaction.
-		txn := kvstore.NewTxn(reader)
-		resp, err = cmd.PrepareWrites(&txn)
-		if err != nil {
-			return nil, err
-		}
-
-		server.Latches.Validate(&txn, latches)
-
-		// Building the transaction succeeded without conflict, write all writes to backing storage.
-		err = server.innerServer.Write(ctxt, txn.Writes)
-		if err != nil {
-			return nil, err
-		}
-	}
-
-	return resp, nil
+// Run runs a transactional command.
+func (server *Server) Run(cmd commands.Command) (interface{}, error) {
+	return commands.RunCommand(cmd, server.innerServer, server.Latches)
 }
 
 // The below functions are Server's gRPC API (implements TikvServer).
 
 // Transactional API.
 func (server *Server) KvGet(_ context.Context, req *kvrpcpb.GetRequest) (*kvrpcpb.GetResponse, error) {
-	cmd := NewGet(req)
+	cmd := commands.NewGet(req)
 	resp, err := server.Run(&cmd)
 	return resp.(*kvrpcpb.GetResponse), err
 }
 
 func (server *Server) KvScan(_ context.Context, req *kvrpcpb.ScanRequest) (*kvrpcpb.ScanResponse, error) {
-	cmd := NewScan(req)
+	cmd := commands.NewScan(req)
 	resp, err := server.Run(&cmd)
 	return resp.(*kvrpcpb.ScanResponse), err
 }
 
 func (server *Server) KvPrewrite(_ context.Context, req *kvrpcpb.PrewriteRequest) (*kvrpcpb.PrewriteResponse, error) {
-	cmd := NewPrewrite(req)
+	cmd := commands.NewPrewrite(req)
 	resp, err := server.Run(&cmd)
 	return resp.(*kvrpcpb.PrewriteResponse), err
 }
 
 func (server *Server) KvCommit(_ context.Context, req *kvrpcpb.CommitRequest) (*kvrpcpb.CommitResponse, error) {
-	cmd := NewCommit(req)
+	cmd := commands.NewCommit(req)
 	resp, err := server.Run(&cmd)
 	return resp.(*kvrpcpb.CommitResponse), err
 }
 
 func (server *Server) KvCheckTxnStatus(_ context.Context, req *kvrpcpb.CheckTxnStatusRequest) (*kvrpcpb.CheckTxnStatusResponse, error) {
-	cmd := NewCheckTxnStatus(req)
+	cmd := commands.NewCheckTxnStatus(req)
 	resp, err := server.Run(&cmd)
 	return resp.(*kvrpcpb.CheckTxnStatusResponse), err
 }
 
 func (server *Server) KvBatchRollback(_ context.Context, req *kvrpcpb.BatchRollbackRequest) (*kvrpcpb.BatchRollbackResponse, error) {
-	cmd := NewRollback(req)
+	cmd := commands.NewRollback(req)
 	resp, err := server.Run(&cmd)
 	return resp.(*kvrpcpb.BatchRollbackResponse), err
 }
 
 func (server *Server) KvResolveLock(_ context.Context, req *kvrpcpb.ResolveLockRequest) (*kvrpcpb.ResolveLockResponse, error) {
-	cmd := NewResolveLock(req)
+	cmd := commands.NewResolveLock(req)
 	resp, err := server.Run(&cmd)
 	return resp.(*kvrpcpb.ResolveLockResponse), err
 }
 
-// Raw API.
+// Raw API. These commands are handled inline rather than by using Run and am implementation of the Commands interface.
+// This is because these commands are fairly straightforward and do not share a lot of code with the transactional
```

Comment:
it should share nothing related code with transactional commands.

Reply:
To clarify, I meant that there is very little code that is the same in both. An example of shared code is that they both use dbreader, but through different abstractions.

Reply:
okay